doc/rst/technotes/errorHandling.rst

```diff
@@ -4,190 +4,581 @@
+  proc catchingErrorsPropagate() throws {
+    try {
+      var x = alwaysThrows(-1);
+      writeln("never reached");
+    } catch e:FileNotFoundError {
+      writeln("caught an error");
+    }
+    // errors other than FileNotFoundError propagate
   }
 
-* If none of the type filters matches the error, a catchall block will be
-  inserted. Its functionality will depend on whether a ``try`` or ``try!``
-  was used.
- 
-  * If ``try`` was used, the error will be propagated. (More on this below.)
+catch-less ``try``
+******************
 
-  * If ``try!`` was used, the program will ``halt()``.
+A logical extension of the above is the case where no ``catch`` blocks
+are attached to the ``try``. Here the ``try`` keyword marks throwing
+calls to clarify control flow.
 
-* ``try`` blocks may be used without any ``catch`` clauses, or without brackets
-  for single statements.
+.. code-block:: chapel
+
+  proc propagatesError() throws {
+    // control flow changes if an error was thrown;
+    // could be indicated more clearly with try
+    canThrow(0);
+
+    try canThrow(0);
+
+    try {
+      canThrow(0);
+    }
+
+    var x = try canThrow(1);
+    writeln(x);
+
+    return try canThrow(0);
+  }
+
+``try`` expressions
+*******************
+
+``try`` and ``try!`` are also available as expressions to clarify
+control flow for assignments and returns. These may not be used
+with ``catch`` clauses.
 
 .. code-block:: chapel
 
-  try {
-    ioSourceOne();
-    ioSourceTwo();
-    ioSourceThree();
+  proc expressionTry(): int throws {
+    var x = try canThrow(1);
+    writeln(x);
+
+    return try canThrow(0);
   }
 
-  try! cannotFail();
-  try! var x = requiredVariable();
+For more information on enforcing clear control flow with the compiler,
+see :ref:`technote-errorHandling-strict` below.
 
+.. _technote-errorHandling-complete:
 
-Propagation
-+++++++++++
-Propagation means that the error will be passed off to a set of handlers
-beyond the current scope. This can be accomplished in two ways:
+Complete handling
++++++++++++++++++
+
+For a function to handle errors from its calls without itself throwing,
+its ``try``/``catch`` must be *complete*. This may be accomplished
+in two ways:
 
-1. Out of the enclosing function, if the function is declared with ``throws``.
+1. A catchall clause on ``try``. This prevents ``try`` from
+propagating the error out of the function as described above.
 
 .. code-block:: chapel
 
-  proc ioSetup() throws {
+  proc warnsOnError(i: int): int {
     try {
-      ioSourceOne();
-      ioSourceTwo();
-      ioSourceThree();
+      alwaysThrows(i);
+    } catch e {
+      writeln("Warning: caught a error ", e);
+    }
+  }
+
+2. ``try!`` instead of ``try``. This will halt the program if no matching
+``catch`` clause is found, instead of propagating.
+
+.. code-block:: chapel
+
+  proc haltsOnError(i: int): int {
+    try! {
+      canThrow(i);
+    } catch e: PrimerError {
+      writeln("caught a PrimerError");
```

Comment:
Another one to rename since this isn't a primer anymore

Reply:
I knocked these out already, changed them to `DemoError`.